cmd/npmExecuteLint.go

```diff
@@ -0,0 +1,139 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/npm"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"path"
+	"strconv"
+	"strings"
+)
+
+func npmExecuteLint(config npmExecuteLintOptions, telemetryData *telemetry.CustomData) {
+	utils := npm.UtilsBundle{}
+
+	err := runNpmExecuteLint(&utils)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runNpmExecuteLint(utils npm.Utils) error {
+	packageJSONFiles, err := npm.FindPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	packagesWithCilint := findPackagesWithCilint(packageJSONFiles, utils)
+
+	if len(packagesWithCilint) > 0 {
+		err = runCiLint(packagesWithCilint, utils)
+		if err != nil {
+			return err
+		}
+	} else {
+		runDefaultLint(utils)
+	}
+
+	return nil
+}
+
+func findPackagesWithCilint(packageJSONFiles []string, utils npm.Utils) []string {
+	var packagesWithCilint []string
+
+	for _, file := range packageJSONFiles {
+		var packageJSON map[string]interface{}
+
+		packageRaw, err := utils.FileRead(file)
+		if err != nil {
+			log.Entry().Warnf("failed to read %s to check for existence of ci-lint script: %v", file, err)
+			continue
+		}
+
+		_ = json.Unmarshal(packageRaw, &packageJSON)
+
+		scripts, ok := packageJSON["scripts"].(map[string]interface{})
+		if ok {
+			_, ok := scripts["ci-lint"].(string)
+			if ok {
+				packagesWithCilint = append(packagesWithCilint, file)
+				log.Entry().Info("Discovered ci-lint script in " + file)
+			}
+		}
+	}
+	return packagesWithCilint
+}
+
+func runCiLint(packagesWithCilint []string, utils npm.Utils) error {
+	execRunner := utils.GetExecRunner()
+	oldWorkingDirectory, err := utils.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current working directory before executing ci-lint scripts: %w", err)
+	}
+
+	for _, packageJSON := range packagesWithCilint {
+		dir := path.Dir(packageJSON)
+		err := utils.Chdir(dir)
+		if err != nil {
+			return fmt.Errorf("failed to change into directory for executing ci-lint script: %w", err)
+		}
+
+		// Ignore possible errors when invoking ci-lint script to not fail the pipeline based on linting results
+		err = execRunner.RunExecutable("npm", "run", "--silent", "ci-lint")
```

Comment:
Same as below. Download cache and registry might not be required, but you might not consume features which are centrally implemented in the npm package by directly using the execRunner.

I think this command is quite similar to npmExecuteScripts for all ci-lint without install ...
Also the npmExecuteScripts could benefit from what you implemented here. It might be got to only run the scripts in these folders where the script is defined as you do it here with findPackagesWithCilint.
Maybe you can bring both a little bit closer

Reply:
I am not entirely sure if I understand your comment. If you mean by npm package, the pkg/npm.go, then I don't see anything I can reuse there, except the handling of npm registries. That is something I am not doing but I guess that's wrong and I have to take it into account in npmExecuteLint as well. And for the download cache I thought it is enough to cal the DownloadCacheUtils.injectDownloadCacheInParameters() function in the groovy wrapper?

The current implementation of npmExecuteScripts actually also directly uses the execRunner which is what I am doing as well. 
So do you mean I should basically move the runNpmExecuteScripts function from npmExecuteScripts into a seperate function in the pkg/npm.go, which then centrally (for both npmExecuteScripts & npmExecuteLint) takes care of executing scripts defined in the package.json?